### PR TITLE
Update ACL Plugin for CakePHP 4.x

### PR DESCRIPTION
**Pull request by nook24**
I have updated the ACL plugin for CakePHP 4.x.

In addition i created an example application which uses _Authentication_, _Authorization_ _ACL_ and _CakePdf_ as reference implementation:

https://github.com/nook24/blood-pressure

### CODE REVIEW

**Comment by nook24** [comment]
I would need some help with the tests please :)

**Comment by garas** [comment]
- Remove PHP 7.1 and earlier from `.travis.yml`.
- Update to `"phpunit/phpunit": "^8.0"` in `composer.json`.
- Check Travis output for further failures.
- Look at `4.x` branches of other plugins like `cakephp/bake` or `cakephp/debug_kit`.

File: src/AclExtras.php [hunk 1]
```diff
@@ -303,7 +303,7 @@ protected function _processPlugins($root, array $plugins = [])
      * @param string $prefix Name of the prefix you are making controllers for.
      * @return array
      */
-    protected function _updateControllers($root, $controllers, $plugin = null, $prefix = null)
+    protected function _updateControllers($root, array $controllers, string $plugin = '', string $prefix = '')
```
Comment (Wouter0100) [inline src/AclExtras.php]: Why not keep `$plugin` and `$prefix` nullable? These were nullable in the past and passing an empty string seems wrong here.

Comment (nook24) [reply src/AclExtras.php]: They get passed to `Inflector::camelize()` which requires to get a string now. I assume in Cake 3 `Inflector::camelize(null)` will return an empty string `''`.

So why passing `null` as default if the application makes an empty string out of it anyway.

File: .travis.yml [hunk 2]
```diff
@@ -20,18 +17,10 @@ env:
...
 
 matrix:
   fast_finish: true
 
-  include:
-    - php: 7.0
-      env: MINIMUMS=1 DEFAULT=1 DB=mysql db_dsn='mysql://travis@0.0.0.0/cakephp_test'
-
-    - php: 7.0
-      env: PHPCS=1 DEFAULT=0
```
Comment (garas) [inline .travis.yml]: Update to PHP 7.2 these `MINIMUMS=1` and `PHPCS=1` jobs instead.

File: .travis.yml [hunk 3]
```diff
@@ -1,9 +1,6 @@
 language: php
 
 php:
-  - 5.6
-  - 7.0
-  - 7.1
   - 7.2
   - 7.3
```
Comment (garas) [inline .travis.yml]: Add `7.4snapshot` here and update line below to `- if [[ $TRAVIS_PHP_VERSION != "7.4snapshot" ]]; then phpenv config-rm xdebug.ini; fi`

Comment (nook24) [reply .travis.yml]: Like so?
```diff
@@ -3,6 +3,7 @@ language: php
 php:
   - 7.2
   - 7.3
+  - 7.4snapshot'

 dist: trusty

@@ -21,8 +22,12 @@ env:
 matrix:
   fast_finish: true

+  include:
+    - php: 7.2
+      env: MINIMUMS=1 DEFAULT=1 DB=mysql db_dsn='mysql://travis@0.0.0.0/cakephp_test'
+
 before_install:
-  - if [[ $TRAVIS_PHP_VERSION != 7.3 ]]; then phpenv config-rm xdebug.ini; fi
+  - if [[ $TRAVIS_PHP_VERSION != "7.4snapshot" ]]; then phpenv config-rm xdebug.ini; fi

   - if [ -n "$GH_TOKEN" ]; then composer config github-oauth.github.com ${GH_TOKEN}; fi;
```

Comment (garas) [reply .travis.yml]: Yes

**Comment by nook24** [comment]
Any ideas on this?
```
[Uncaught InvalidArgumentException in /home/travis/build/cakephp/acl/vendor/cakephp/cakephp/src/Database/Driver/Sqlite.php on line 71]

InvalidArgumentException: The `database` key for the `test` SQLite connection needs to be a non-empty string. in /home/travis/build/cakephp/acl/vendor/cakephp/cakephp/src/Database/Driver/Sqlite.php:71
```

File: .travis.yml (hunk 3 above)
Comment (garas) [reply .travis.yml]: Except should have opening quote here `- '7.4snapshot'`...

Overall, `.travis.yml` in https://github.com/cakephp/acl/pull/159/files should be minimally changed, only lower/upper PHP versions changed

Comment (nook24) [reply .travis.yml]: Could you please post a complete travis.yml?

**Comment by garas** [comment]
I think problem is `sqlite::memory:` should be `sqlite:///:memory:` and `db_class` removed in `tests/bootstrap.php` and `phpunit.xml.dist`.

There are many outdated things in this repo. I'm comparing with `4.x` branches of `cakephp/cakephp` and `cakephp/bake` for changes in `tests/bootstrap.php` and `phpunit.xml.dist`.

File: .travis.yml (hunk 2 above)
Comment (garas) [reply .travis.yml]: This `PHPCS=1 DEFAULT=0` job not yet restored

File: tests/bootstrap.php [hunk 4]
```diff
@@ -96,21 +100,12 @@
...
-    'driver' => getenv('db_class'),
-    'dsn' => getenv('db_dsn'),
-    'database' => getenv('db_database'),
-    'username' => getenv('db_login'),
-    'password' => getenv('db_password'),
-    'timezone' => 'UTC',
-    'quoteIdentifiers' => getenv('quoteIdentifiers'),
-]);
+ConnectionManager::setConfig('test', ['url' => getenv('db_dsn')]);
+ConnectionManager::setConfig('test_custom_i18n_datasource', ['url' => getenv('db_dsn')]);
```
Comment (garas) [inline tests/bootstrap.php]: There was no `test_custom_i18n_datasource` before.

Comment (garas) [reply tests/bootstrap.php]: In the end of this file replace `Carbon/Chronos` lines with single `Chronos::setTestNow(Chronos::now());`.

Remove PHPUnit aliases.

Comment (nook24) [reply tests/bootstrap.php]: My issue is, that the tests expects data which is not defined in the fixtures. Like this:
```php
        $result = $Aco->node('controllers/Admin/Posts')->toArray();
        $this->assertEquals($result[0]['alias'], 'Posts');
        $result = $Aco->find('children', ['for' => $result[0]['id']])->toArray();
        $this->assertEquals(count($result), 3);
```
What am i missing here?

Comment (garas) [reply tests/bootstrap.php]: I don't use this plugin so I can't help much on plugin functionality itself.

Code above these lines might be broken (`$this->Task->acoUpdate();` maybe).

I can help on generic 3 to 4 upgrade. There are [many breaking](https://book.cakephp.org/4/en/appendices/4-0-migration-guide.html) changes in 4.0, you might need to fix them first. Use step by step with debugger if necessary.

File: composer.json [hunk 5]
```diff
@@ -21,12 +21,12 @@
         "source": "https://github.com/cakephp/acl"
     },
     "require": {
-        "cakephp/cakephp": "^3.6.0",
+        "cakephp/cakephp": "4.x-dev as 4.0.0",
```
Comment (dereuromark) [inline composer.json]: Can you also add `php >= 7.2` as we usually always do?

File: composer.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "cakephp/acl",
-    "description": "Acl Plugin for CakePHP 3.x framework",
+    "description": "Acl Plugin for CakePHP 4.x framework",
```
Comment (dereuromark) [inline composer.json]: Please drop version here completely. This should never be in the descr.

File: src/Auth/CrudAuthorize.php [hunk 7]
```diff
@@ -51,7 +51,7 @@ public function __construct(ComponentRegistry $registry, $config = [])
      * @param \Cake\Network\Request $request The request needing authorization.
      * @return bool
      */
-    public function authorize($user, ServerRequest $request)
+    public function authorize($user, ServerRequest $request) :bool
```
Comment (dereuromark) [inline src/Auth/CrudAuthorize.php]: CS issue
`): bool` etc

File: src/AclExtras.php (hunk 1 above)
Comment (dereuromark) [reply src/AclExtras.php]: You can just check for null before inflecting probably.
Might be a bit cleaner then.

**Comment by dakota** [comment]
@nook24 Any movement on this?

**Comment by nook24** [comment]
Nope, I don't have time to look into the tests or code style stuff yet.

How ever, the plugin works without any issues. Any help is welcome to get the merged :)

**Comment by dereuromark** [comment]
The beta release is missing for https://github.com/FriendsOfCake/awesome-cakephp#authentication-and-authorization update.

**Comment by rchavik** [comment]
I'm going to try to merge this

**Comment by rchavik** [comment]
Merged and updated tests.  Completed with warning in my local env with 4.0.  Failed with error in travis with cake 4.1.

Will stop here now.

**Comment by nook24** [comment]
I just can confirm the the Plugin itself run on CakePHP 4.1 and PHP 7.2, 7.3 and 7.4 ^^

Unfortunately I can't find your Trevis output

**Comment by rchavik** [comment]
> I just can confirm the the Plugin itself run on CakePHP 4.1 and PHP 7.2, 7.3 and 7.4 ^^
> 
> Unfortunately I can't find your Trevis output

Last run for 4.x was: https://travis-ci.org/github/cakephp/acl/builds/724969886.  Managed to get it all green.

**Pull request merged**